pygmt/src/triangulate.py

```diff
@@ -0,0 +1,384 @@
+"""
+triangulate - Delaunay triangulation or Voronoi partitioning and gridding of
+Cartesian data.
+"""
+import warnings
+
+import pandas as pd
+from pygmt.clib import Session
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import (
+    GMTTempFile,
+    build_arg_string,
+    fmt_docstring,
+    kwargs_to_strings,
+    use_alias,
+)
+from pygmt.io import load_dataarray
+
+
+class triangulate:  # pylint: disable=invalid-name
+    """
+    Delaunay triangulation or Voronoi partitioning and gridding of Cartesian
+    data.
+
+    Triangulate reads in x,y[,z] data and performs Delaunay triangulation,
+    i.e., it finds how the points should be connected to give the most
+    equilateral triangulation possible. If a map projection (give *region* and
+    *projection*) is chosen then it is applied before the triangulation is
+    calculated. By default, the output is triplets of point id numbers that
+    make up each triangle. The id numbers refer to the points position (line
+    number, starting at 0 for the first line) in the input file. If **outgrid**
+    and **spacing** are set a grid will be calculated based on the surface
+    defined by the planar triangles. The actual algorithm used in the
+    triangulations is either that of Watson [1982] or Shewchuk [1996] [Default]
+    (if installed; type **gmt get GMT_TRIANGULATE** on the command line to see
+    which method is selected). This choice is made during the GMT installation.
+    Furthermore, if the Shewchuk algorithm is installed then you can also
+    perform the calculation of Voronoi polygons and optionally grid your data
+    via the natural nearest neighbor algorithm. **Note**: For geographic data
+    with global or very large extent you should consider
+    :gmt-docs:`sphtriangulate <sphtriangulate.html>` instead since
+    **triangulate** is a Cartesian or small-geographic area operator and is
+    unaware of periodic or polar boundary conditions.
+    """
+
+    @staticmethod
+    @fmt_docstring
+    @use_alias(
+        G="outgrid",
+        I="spacing",
+        J="projection",
+        R="region",
+        V="verbose",
+        b="binary",
+        d="nodata",
+        e="find",
+        f="coltypes",
+        h="header",
+        i="incols",
+        r="registration",
+        s="skiprows",
+        w="wrap",
+    )
+    @kwargs_to_strings(I="sequence", R="sequence", i="sequence_comma")
+    def _triangulate(
+        data=None, x=None, y=None, z=None, output_type=None, outfile=None, **kwargs
```

Comment:
```suggestion
        data=None, x=None, y=None, z=None, *, output_type, outfile=None, **kwargs
```
I think that output type should be required to prevent unexpected behavior.

Reply:
I saw this in your `grdhisteq` implementation and was wondering what this is about. Is this some new Python syntax? I don't think I've seen it very often.

Reply:
I never really used these notations for specifying positional-only, positional-or-keyword, or keyword-only before doing a bunch of research about these specifications when trying to figure out how to get tab completion to work with `@use_alias`. I found the 3 PEPs listed below helpful. In this case, `output_type` would need to be specified as keyword-only to be required since it comes after optional positional-or-keywords arguments.

I agree it's not used very often and would be interested in knowing why that's the case.

https://peps.python.org/pep-0457/
https://peps.python.org/pep-0570/
https://peps.python.org/pep-3102/